packages/headless/src/controllers/facets/range-facet/numeric-facet/headless-numeric-filter.ts

```diff
@@ -0,0 +1,207 @@
+   *
+   * Minimum: `0`
+   *
+   * @defaultValue `1000`
+   */
+  injectionDepth?: number;
+}
+
+export interface NumericFilterRange {
+  /**
+   * The starting value for the numeric range.
+   */
+  start: number;
+
+  /**
+   * The ending value for the numeric range.
+   */
+  end: number;
+}
+
+export interface NumericFilterProps {
+  /**
+   * The options for the `NumericFilter` controller.
+   */
+  options: NumericFilterOptions;
+  /**
+   * The initial range that should be applied to the `NumericFilter` controller.
+   */
+  initialState?: NumericFilterRange;
+}
+
+export interface NumericFilterState {
+  /**
+   * The facet ID.
+   * */
+  facetId: string;
+
+  /**
+   * The current selected range.
+   */
+  range?: NumericFacetValue;
+
+  /**
+   * `true` if a search is in progress and `false` otherwise.
+   */
+  isLoading: boolean;
+}
+
+/**
+ * The `NumericFilter` controller makes it possible to create a numeric filter.
+ */
+export interface NumericFilter extends Controller {
+  /**
+   * Deselects current filter.
+   */
+  deselect(): void;
+
+  /**
+   * Updates the selected range.
+   *
+   * @param range - The numeric range.
+   */
+  setRange(range: NumericFilterRange): void;
+
+  /**
+   * The state of the `NumericFacet` controller.
+   */
+  state: NumericFilterState;
+}
+
+export function buildNumericFilter(
+  engine: SearchEngine,
+  props: NumericFilterProps
+): NumericFilter {
+  if (!loadNumericFilterReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const facetId = determineFacetId(engine, props.options);
+  const options: RegisterNumericFacetActionCreatorPayload = {
+    ...props.options,
+    currentValues: props.initialState
+      ? [{...props.initialState, endInclusive: true, state: 'selected'}]
+      : [],
+    generateAutomaticRanges: false,
+    facetId,
+  };
+
+  validateNumericFacetOptions(engine, options);
+  dispatch(registerNumericFacet(options));
+
+  return {
+    ...controller,
+    deselect: () => {
+      dispatch(deselectAllFacetValues(facetId));
+      dispatch(updateFacetOptions({freezeFacetOrder: true}));
+      dispatch(executeSearch(logFacetClearAll(facetId)));
+    },
+    setRange: (range) => {
+      const facetValue: NumericFacetValue = {
+        ...range,
+        state: 'selected',
+        numberOfResults: 0,
+        endInclusive: true,
+      };
+
+      dispatch(
+        updateNumericFacetValues({
+          facetId,
+          values: [facetValue],
+        })
+      );
+      dispatch(updateFacetOptions({freezeFacetOrder: true}));
+      dispatch(
+        executeSearch(
+          logFacetSelect({
+            facetId,
+            facetValue: `${facetValue.start}..${facetValue.end}`,
+          })
+        )
+      );
+    },
+
+    get state() {
+      const isLoading = engine.state.search.isLoading;
+      const response = baseFacetResponseSelector(engine.state, facetId) as
```

Comment:
I think you could leverage `isNumericFacetResponse` from `numeric-facet-selector.ts` instead of casting.